ivadomed/postprocessing.py

```diff
@@ -41,37 +41,47 @@ def threshold_predictions(predictions, thr=0.5):
 def keep_largest_object(predictions):
     """
     Keep the largest connected object from the input array (2D or 3D).
-    Note: This function only works for binary segmentation.
+    Note: If the input is not binary, the function calls the thresholding with low value (here 10e-3),
+        applies morphomath operation, and then uses the mask_prediction function to apply the operation
+        on the soft pred based on the binary output of the morphomath process.
 
     Args:
-        predictions (array or nibabel object): Input binary segmentation. Image could be 2D or 3D.
+        predictions (array or nibabel object): Input segmentation.
+            Image could be 2D or 3D, soft or binary.
     Returns:
         Array or nibabel (same object as the input).
     """
-    assert np.array_equal(predictions, predictions.astype(bool))
     predictions_proc = np.copy(predictions)
+    # If input is not binary, then make it binary by thresholding it
+    if not np.array_equal(predictions, predictions.astype(bool)):
```

Comment:
I believe you can use the same trick as for the @nifti_capable to minimize code duplication. See my suggestions for `keep_largest_object()`, which you can apply to the rest of the functions if you agree with the suggestions: 
https://github.com/neuropoly/ivado-medical-imaging/compare/cg/postpro-soft-friendly...jca/postpro-soft-friendly

Reply:
(Oups... I have just seen this comment)
Very elegant solution! Thank you very much, I am merging these